### ISSUE TITLE
streamUI doesn't support multiple parallel tool calls

### ISSUE THREAD

**Issue by mfclarke-cnx** [vercel/ai#1894]
### Feature Description

Allow streamUI to gracefully handle multiple parallel tool calls. Currently when the model makes multiple tool calls, the tool handler gets called multiple times, but the first call to resolve closes the UI steam and the rest throw:

```
⨯ node_modules/.pnpm/ai@3.1.26_openai@4.48.1_react@18.3.1_solid-js@1.8.17_svelte@4.2.17_vue@3.4.27_zod@3.23.8/node_modules/ai/rsc/dist/rsc-server.mjs (1231:12) @ assertStream
 ⨯ unhandledRejection: Error: .update(): UI stream is already closed.
 ```

### Use Case

Any feature where the model can call multiple tools in parallel. For example, creating multiple things, searching multiple websites, updating multiple things.

**Comment by camwest** [comment]
It looks like this hardcoded "true" is where the error is happening.

https://github.com/vercel/ai/blob/f7002ad2c5aa58ce6ed83e8d31fe22f71ebdb7d7/packages/core/rsc/stream-ui/stream-ui.tsx#L303

It's a bad assumption. The tool-call event doesn't immediately mean it's the last render call because there may be multiple tool-call events.

**Comment by camwest** [comment]
A potential fix would be to add a 'all-tool-calls-done' event which gets pushed to stream after the loop right around here:

https://github.com/vercel/ai/blob/f7002ad2c5aa58ce6ed83e8d31fe22f71ebdb7d7/packages/openai/src/openai-chat-language-model.ts#L343

Then we handle that in stream-ui and finalize the stream there.

The challenge is we'd have to update anything that implements LanguageModelV1 to trigger this event, or version to V2.

**Comment by camwest** [comment]
Another option is to close the stream on finish:

https://github.com/vercel/ai/blob/f7002ad2c5aa58ce6ed83e8d31fe22f71ebdb7d7/packages/openai/src/openai-chat-language-model.ts#L382-L388

Currently not implemented in streamUI:

https://github.com/vercel/ai/blob/f7002ad2c5aa58ce6ed83e8d31fe22f71ebdb7d7/packages/core/rsc/stream-ui/stream-ui.tsx#L314

**Comment by bneigher** [comment]
ah this is the culprit of my issue! been doing this since `v3.1.25`

**Comment by camwest** [comment]
@mfclarke-cnx can you change this to a bug instead of enhancement?

**Comment by mfclarke-cnx** [comment]
@camwest unfortunately I don't have permissions to change the labels. @lgrammel would you be able to update labels please?

**Comment by miguelvictor** [comment]
I also have this issue but the weird thing about mine is the same tool is being called twice.

**Comment by gclark-eightfold** [comment]
As tagged by @bneigher, I have an approach I'm working on in #1904, but it's still a WIP.

Main issue I have right now is that because we don't know how many tool calls will be made ahead of time, the parent "streamableUI" doesn't call `.done` until all tool calls complete.

I think my expectation is that the UI will be added to as the final tool calls stream in.

**Comment by camwest** [comment]
> Main issue I have right now is that because we don't know how many tool calls will be made ahead of time, the parent "streamableUI" doesn't call `.done` until all tool calls complete.

I think you can get this information at the provider level. 

For example the OpenAI provider passes an array of tool_calls so you know when you're finished iterating that it's 'done'

**Comment by HECHONG999** [comment]
"Parallel support, how to resolve the 'done' error bug now."

**Comment by dutsik-p** [comment]
Maybe anybody knows how to tune the prompt so the model doesnot put number of tools to call at the same time?

**Comment by grrowl** [comment]
As a kind of "workaround" you can set `parallelToolCalls` to `false` (as introduced in #1937) — but lack of parallel tool calling support really should be mentioned on https://sdk.vercel.ai/docs/troubleshooting/common-issues

Frustrating that this common benefit of AI tool calling is broken on Vercel AI UI :(

**Comment by aayush-kapoor** [comment]
Marking this as closed for now - if you face any regression or have a similar problem after migrating to the latest AI SDK version, please feel free to open a new issue (with possible reproduction)

**Comment by lgrammel** [comment]
## Identify issue type

Identified as **Bug**  
Confidence: **High**

### Rationale
This reports a concrete runtime failure in existing `streamUI` behavior when a model makes parallel tool calls. Although fixes and documentation are discussed, the primary issue is an incorrectly closed UI stream.

### Evidence
- Multiple parallel tool calls invoke their handlers, but the first resolution closes the UI stream and subsequent calls throw `.update(): UI stream is already closed`.
- Comments identify premature stream finalization in `streamUI` and describe the issue as occurring since v3.1.25.
- Disabling parallel tool calls is cited only as a workaround; the expected parallel behavior remains broken.
